pyparsing/core.py

```diff
@@ -645,20 +605,21 @@ def _parseNoCache(self, instring, loc, doActions=True, callPreParse=True):
                             and isinstance(tokens, (ParseResults, list)),
                             modal=self.modalResults,
                         )
-        if debugging:
-            # ~ print("Matched", self, "->", retTokens.asList())
-            if self.debugActions[MATCH]:
-                self.debugActions[MATCH](instring, tokensStart, loc, self, retTokens)
+            except Exception as err:
+                self.debug and self.debugActions.FAIL(instring, start, self, err)
+                raise
+        self.debug and self.debugActions.MATCH(instring, start, loc, self, retTokens)
 
         return loc, retTokens
 
     def tryParse(self, instring, loc, raise_fatal=False):
-        try:
-            return self._parse(instring, loc, doActions=False)[0]
-        except ParseFatalException:
-            if raise_fatal:
-                raise
-            raise ParseException(instring, loc, self.errmsg, self)
+        with ParserElement.packrat_cache.lock:
```

Comment:
Instead of locking upon every `_parseCache` call, we lock it once when we enter into the parsing routine.

Reply:
What if packrat is not enabled? Would these unneeded lock/unlocks impact parse-time performance?

Reply:
Good point.  We need some stress tests to get better feedback about if changes improve performance or not.  The current test suite is not a good example because it is spending most of its time constructing grammars, not actually parsing.  If you know of some large examples, I can turn them into a stress test, with timing reporting, intended to be run manually.

The speed impact of this lock does depend on use-cases:  If there is a single large parse event, then this is small cost (and much smaller when packrat enabled).  If the user is calling `parseString` (or `tryParse`) many times on small input, then the cost of this lock is larger.  I do not know how often this latter scenario happens.

Even so, there was some discussion about parsing with multiple threads. This lock helps with that: allowing only one thing to be parsed at a time.

I am not familiar with all the entry points; I will need your guidance on that.  Maybe I will make 

```
@entrypoint
def parseString(...
```

That will perform the lock, and mark the entry points.